sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala

```diff
@@ -59,6 +59,30 @@ trait PredicateHelper {
     }
   }
 
+  /**
+   * Convert an expression into its conjunctive normal form (CNF), i.e. AND of ORs.
+   * For example, a && b || c is normalized to (a || c) && (b || c) by this method.
+   *
+   * Refer to https://en.wikipedia.org/wiki/Conjunctive_normal_form for more information
+   */
+  protected def cnfNormalization(condition: Expression): Expression = {
+    condition.transformUp {
+      case or: Or => pushOrToBottom(or)
+    }
+  }
+
+  private def pushOrToBottom(condition: Expression): Expression = {
+    condition match {
+      case Or(left, right) => (left, right) match {
+        case (And(ll, lr), r) =>
+          And(pushOrToBottom(Or(ll, r)), pushOrToBottom(Or(lr, r)))
+        case (l, And(rl, rr)) =>
+          And(pushOrToBottom(Or(l, rl)), pushOrToBottom(Or(l, rr)))
+        case (l, r) => Or(l, r)
+      }
+    }
+  }
```

Comment:
The following version may be questionably more readable:

```scala
  private def pushOrToBottom(condition: Expression): Expression = {
    condition match {
      case Or(And(innerLhs, innerRhs), rhs) =>
        And(pushOrToBottom(Or(innerLhs, rhs)), pushOrToBottom(Or(innerRhs, rhs)))

      case Or(lhs, And(innerLhs, innerRhs)) =>
        And(pushOrToBottom(Or(lhs, innerLhs)), pushOrToBottom(Or(lhs, innerRhs)))

      case _ => condition
    }
  }
```

Shall we also cover cases like `Not(Or(x, y))` => `And(Not(x), Not(y))` here?